xarray/tests/duckarrays/base/strategies.py

```diff
@@ -0,0 +1,199 @@
+from typing import Any
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+
+import xarray as xr
+from xarray.core.utils import is_dict_like
+
+from . import utils
+
+all_dtypes = (
+    npst.integer_dtypes()
+    | npst.unsigned_integer_dtypes()
+    | npst.floating_dtypes()
+    | npst.complex_number_dtypes()
+)
+
+
+def numpy_array(shape, dtypes=None):
+    if dtypes is None:
+        dtypes = all_dtypes
+
+    def elements(dtype):
+        max_value = 100
+        min_value = 0 if dtype.kind == "u" else -max_value
+
+        return npst.from_dtype(
+            dtype, allow_infinity=False, min_value=min_value, max_value=max_value
+        )
+
+    return dtypes.flatmap(
+        lambda dtype: npst.arrays(dtype=dtype, shape=shape, elements=elements(dtype))
+    )
+
+
+def dimension_sizes(min_dims, max_dims, min_size, max_size):
+    sizes = st.lists(
+        elements=st.tuples(st.text(min_size=1), st.integers(min_size, max_size)),
+        min_size=min_dims,
+        max_size=max_dims,
+        unique_by=lambda x: x[0],
+    )
+    return sizes
+
+
+# Is there a way to do this in general?
+# Could make a Protocol...
+T_DuckArray = Any
```

Comment:
I added one or two type hints just as code documentation for myself, but I'm not sure if typing these base classes in general would yield any benefit in terms of catching errors with duck array integrations?

Reply:
Using a protocol (see @Illviljan 's [links](https://github.com/pydata/xarray/issues/6894#issuecomment-1209756650)) would be perfect in theory but in practice xarray doesn't actually require all the attributes defined in the array API, only a subset, so perhaps we should maintain our own duck array protocol so as not to be overly strict...?

Reply:
I think you should move `T_Duckarray `to `xarray.core.types` and then import it from there. It should be useful in other places as well.

Yeah xarray may not use everything available in the array api (yet). But as you've noticed we may get a lot of hypothesis and typing tests figured out for us.